### ISSUE TITLE
[Python] Implement new Concept Exercise: Enum

### ISSUE THREAD

**Issue by BethanyG** [exercism/v3#2309]
This issue describes how to implement the `Enum` concept exercise for the Python track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student how `Enums` ([_Enumerations_](https://en.wikipedia.org/wiki/Enumerated_type)) are implemented & used in Python. We will teach this through Pythons `Enum` class/type.

## Learning objectives

*  What is an `enumeration` / `enum` is, and why they are needed/wanted in Python?
*  Understand the nomenclature (_naming_) used in reference to `enums` (**e.g.** `enumeration`/`enum`, `enum members`, enum member _`names`_, and enum member _`values`_)
*  Understand that `enumeration members` are functionally `constants` (_[and therefore should be formatted as such](https://www.python.org/dev/peps/pep-0008/#constants)_)
*  How [`Enums` are different](https://docs.python.org/3/library/enum.html#how-are-enums-different) from other, more generic Python classes.

*  How to create various `Enums` 
    *  By using the `class` syntax by importing & subclassing `Enum`.
    *  By using the [Enum Functional API](https://docs.python.org/3/library/enum.html#functional-api)
    *  By defining `enum members` with `values` that  include _non-int_ types (_like `str`, `tuple`,  `float` etc._)  
    *  Using the `auto()` function to automatically assign integer values to members when exact `value` types are unimportant/not needed.
    * Creating  member _aliases_ by assigning the same `value` to different `names` (_two different `names` can have the same `value`; the second name defined becomes an **alias** of the first_)
     *  Using the class decorator `@enum.unique` to prevent member aliases and enforce that member `values` be unique.

*  How to use an `enumeration` and `enumeration members` in other functions/code
    *  `Enum members` are _iterable_ in member definition order, but iteration **will not include aliases**.
    *  An ordered mapping of names to members can be retrieved via `__members__.items()`
    *  `enumeration members` are compared by **identity**, using the `is`/`is not` keywords
    *  [Ordered comparison](https://docs.python.org/3/reference/expressions.html#value-comparisons) (`<`, `>`, `<=`, '>=`) between `enumeration values` is not supported, and will throw a `TypeError`.  
    *  Equality/inequality comparison is defined, and  `==` and  `!=` can be used.
    *  Comparisons between `enumeration values`  and `non-enumeration values` will always return `False`  

## Out of scope

* [Flag]() `enum` subtype (_perhaps better as an advanced exercise that includes bitwise operations_)
* [IntEnum and IntFlag]() subtypes (_since they break the semantic promises of an `enum` by being comparable to `int`_)
* mixins & multiple inheritance for the creation of `Enums`
* using [`__new__()` to customize the value of an `enum` member](https://docs.python.org/3/library/enum.html#when-to-use-new-vs-init) (_better for an advanced/extended enum exercise_)
* [omitting values](https://docs.python.org/3/library/enum.html#omitting-values)
* subclassing an "empty" pre-defined `enum`
* customization of `auto()`
* Pickling

## Concepts

- enumeration, enums

## Prerequisites

- `decorators`, `@`
- `__init__()`
- `classes`, `OOP`
- `inheritance`
- `iteration`
- `iterables`
- `dunder methods`
- `comparisons`
- `rich comparisons`
- `class attributes`
- `importing`
- `aliasing`
- `dicts`, `dict methods` (specifically `dict.items()`)
- `mapping types`
- `immutable`,` immutability`
- `class properties`

## Resources to refer to

- [Exercism v3 C# `Enums` concept exercise](https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/enums)
- [Python Docs: `Enum`](https://docs.python.org/3/library/enum.html)
- [Enum - Mouse vs. Python](https://www.blog.pythonlibrary.org/2018/03/20/python-3-an-intro-to-enumerations/)
- [Why you should use more enums in Python - Florian Dahlitz ](https://florian-dahlitz.de/blog/why-you-should-use-more-enums-in-python)
- [Python Docs: How are Enums Different?](https://docs.python.org/3/library/enum.html#how-are-enums-different)
- [Python Docs: The Enum Functional API](https://docs.python.org/3/library/enum.html#functional-api)
- [Stack Overflow: Python enum, when and where to use?](https://stackoverflow.com/questions/22586895/python-enum-when-and-where-to-use)
- [PEP435: Adding an Enum Type to the Python Standard Library](https://www.python.org/dev/peps/pep-0435/)
- [Using Enums and Django model Choices - Ben Cleary](https://medium.com/@bencleary/using-enums-as-django-model-choices-96c4cbb78b2e)

### Hints

Should be focused on basic definition & use, and refer to one or more of the Python related links above (_or equivalent/analogous links_).

### After

Some links and extended discussion of what problems an enumeration type is trying to solve in python programming might be useful to the student here.  In particular, the rejected [PEP354](https://www.python.org/dev/peps/pep-0354/#rejection-notice), and the subsequent accepted [PEP435](https://www.python.org/dev/peps/pep-0435/#motivation), as well as other discussion and links from the resources section.

It might also be useful to point the student toward all the rich customization options available for `enum`, such as the ["interesting examples"](https://docs.python.org/3/library/enum.html#interesting-examples) section of the docs.

Finally, some discussion, links and usage  around the "out of scope" cases above could be great to encourage the student to explore `enums` even further.

## Representer

TBD

## Analyzer

TBD

## Implementing

Tests should be written using unittest.TestCase (_we try to avoid using PyTest specific features_), & the test file shoul dbe named `enum_test.py`.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue, and/or reach out on the `#python-maintainers` channel in the exercism Slack team.

**Comment by BethanyG** [comment]
@mohanrajanr - all yours!  LMK if you have questions or issues.

**Comment by BethanyG** [comment]
@cmccandless  - when you get a chance, if you could give this a quick "sniff test" as a concept/issue & LMK if we need to edit it that would be **awesome**.  _Thank you_.

**Comment by mohanrajanr** [comment]
Link to the PR Here: https://github.com/exercism/v3/pull/2312

**Comment by cmccandless** [comment]
> @cmccandless - when you get a chance, if you could give this a quick "sniff test" as a concept/issue & LMK if we need to edit it that would be awesome. Thank you.

My thoughts:
- `Learning objectives` section feels like it's too much for a single exercise. Maybe move Functional API and aliases/`@unique` to "Out of scope"?
- Perhaps a mention in after.md (or whatever document is replacing it) about how `enumerate` is not the same thing as `Enum`; someone new to Python might learn `Enum`, then see `enumerate` and get confused.

**Comment by BethanyG** [comment]
Excellent feedback, @cmccandless .  Thank you.  I went ahead and merged the PR, but will be filing some improvement issues as I go back through our existing exercises and fill in concept docs, re-title, etc. etc.  Makes sense to move `@unique` out, as well as aliases.  I think the `Functional API`, being an alternative and shorter constructor should stay.  

And we definitely want  the `Enum` vs `enumerate()` distinction in the `after.md` or `approaches.md` .

**Comment by BethanyG** [comment]
PR  #2312  Has been merged.  Closing this issue.
